marketplace-service/src/main/java/com/axonivy/market/service/impl/MetadataServiceImpl.java

```diff
@@ -0,0 +1,331 @@
+    }
+  }
+
+  private String updateImagesWithDownloadUrl(Product product, String unzippedFolderPath,
+      String readmeContents) throws IOException {
+    List<Path> allImagePaths = Files.walk(Paths.get(unzippedFolderPath))
+        .filter(Files::isRegularFile)
+        .filter(path -> path.getFileName().toString().toLowerCase().matches(CommonConstants.IMAGE_EXTENSION))
+        .toList();
+
+    Map<String, String> imageUrls = new HashMap<>();
+    allImagePaths.forEach(imagePath -> Optional.of(imageService.mappingImageFromDownloadedFolder(product, imagePath))
+        .ifPresent(image -> imageUrls.put(imagePath.getFileName().toString(),
+            CommonConstants.IMAGE_ID_PREFIX.concat(image.getId()))));
+
+    ProductContentUtils.replaceImageDirWithImageCustomId(imageUrls, readmeContents);
+
+    return readmeContents;
+  }
+
+  private String extractProductJsonContent(Path filePath) {
+    try {
+      InputStream contentStream = MavenUtils.extractedContentStream(filePath);
+      return IOUtils.toString(contentStream, StandardCharsets.UTF_8);
+    } catch (Exception e) {
+      log.error("Cannot extract product.json file {}", e.getMessage());
+      return null;
+    }
+  }
+
+  public void updateMavenArtifactVersionFromMetadata(MavenArtifactVersion artifactVersionCache,
+      Metadata metadata) {
+    NonStandardProduct currentProduct = NonStandardProduct.findById(metadata.getProductId());
+    metadata.getVersions().forEach(version -> {
+      if (VersionUtils.isSnapshotVersion(version) && currentProduct != NonStandardProduct.PORTAL) {
+        if (VersionUtils.isOfficialVersionOrUnReleasedDevVersion(metadata.getVersions().stream().toList(), version)) {
+          updateMavenArtifactVersionForNonReleaseDeVersion(artifactVersionCache, metadata, version);
+        }
+      } else {
+        updateMavenArtifactVersionCacheWithModel(artifactVersionCache, version, metadata);
+      }
+    });
+  }
+
+  public void updateMavenArtifactVersionForNonReleaseDeVersion(MavenArtifactVersion artifactVersionCache,
+      Metadata metadata, String version) {
+    Metadata snapShotMetadata = MavenUtils.buildSnapShotMetadataFromVersion(metadata, version);
+    MetadataReaderUtils.parseMetadataSnapshotFromString(MavenUtils.getMetadataContentFromUrl(snapShotMetadata.getUrl()),
+        snapShotMetadata);
+    updateMavenArtifactVersionCacheWithModel(artifactVersionCache, version, snapShotMetadata);
+  }
+
+  public void updateArtifactsFromNonSyncedVersion(String productId, List<String> nonSyncedVersions,
+      Set<Artifact> artifacts) {
+    if (CollectionUtils.isEmpty(nonSyncedVersions)) {
+      return;
+    }
+    nonSyncedVersions.forEach(version -> {
+      ProductJsonContent productJson = productJsonRepo.findByProductIdAndVersion(productId, version);
+      List<Artifact> artifactsInVersion = MavenUtils.getMavenArtifactsFromProductJson(productJson);
+      artifacts.addAll(artifactsInVersion);
```

Comment:
set artifacts as a return object, then it that call function, use that return value, because you just new artifacts, then pass to this function,